### ISSUE TITLE
thread_ctrl::wait_for behaviour

### ISSUE THREAD

**Issue by jjsat** [RPCS3/rpcs3#4031]
I touched on that in #3918, but from my understanding (which could be wrong :-) there is an issue with thread_ctrL::wait_for. It is used throughout the emulator where the current thread is supposed to be delayed, for example while displaying a save dialog or a message dialog. But from my experience the behaviour is somewhat random.

For example Wolfenstein The New Order displays a message dialog during installaion. During that the background is animated and the sound continues playing. This matches the PS3 behaviour, so that is ok.

But for examle in Silent HiIl HD Collection when displaying a save dialog, the screen fades to red and a special sound plays. Both the animation and the sound are abruptly cut off when the dialog appears and only continue after it closes. The PS3 would instead smoothly fade the sound to a background noise that plays while the dialog is open and fade to another background image.

When replacing the waiting code in savedata_op() with:
```
// Display Save Data List asynchronously in the GUI thread.
atomic_t<bool> dlg_result(false);

Emu.CallAfter([&]()
{
	selected = Emu.GetCallbacks().get_save_dialog()->ShowSaveDataList(save_entries, focused, operation, listSet);
	dlg_result = true;
	lv2_obj::awake(ppu);
});

lv2_obj::sleep(ppu);

while (!ppu.state.test_and_reset(cpu_flag::signal))
{
	thread_ctrl::wait();
}
```
the game seems to behave the intended way.

I am not saying that this is the right way, but I wonder if thread_ctrl::wait_for is really working as intended.

With #4011 the save dialog also shows the blocking behaviour in SH, but there the UI code runs its own input loop and blocks the thread with that.

**Comment by kd-11** [comment]
Afaik wait_for makes no guarantees other than that the function will wait for upto whatever delay is requested. If you need guaranteed waiting use thread::sleep_for(). wait_for is intended to only slightly interfere with the calling thread, maybe by a few instructions where sleep_for is not suitable as the minimum resolution is too large. thread::yield is also a good alternative if you want to intentionally trigger the OS scheduler.

**Comment by jjsat** [comment]
I guess the issue also might be with how the method is currently used. Also for the new UI maybe the rendering and button checing could be in its own thread - while the PPU thread that called the dialog sleeps - to avoid the described blocking behaviour.

**Comment by kd-11** [comment]
rendering and button checking is in a separate thread for the new UI. In early code the button check was in the same thread but I don't see how wait_for could have helped there.
As for usage of wait_for, any particular case where it would cause significant slowdown? I haven't written much of rpcs3 code outside of rsx and spu so I don't know how much its used.

**Comment by jjsat** [comment]
Ok. I made comparison videos to show what I mean.

This is Dante's Inferno with the savedialog in ui_overlay branch. Everyhting cuts out when the dialog opens.
[https://www.youtube.com/watch?v=9YL2NBUm8mw](url)

This is with the change in https://github.com/jjsat/rpcs3/commit/be5ad69069c16b66a85e822ac3020b0c968e7fe5 . The sound keeps playing. This is also true for the fallback Qt based dialog.
[https://www.youtube.com/watch?v=dXIGK-dOSEk](url)

I suspect that this could have a performance impact when using thread_ctrl::wait_for elsewhere since it blocks the thread like a blocking function call would, which also keeps the thread scheduler from working correctly.

**Comment by kd-11** [comment]
I should clarify a bit more:
1. thread_ctrl::wait_for isn't a replacement for thread::sleep_for() or busy_wait(); your observation is correct, it need not be used to delay execution arbitrarily. wait_for waits for a signal notification or aborts if n usecs expire before said notification is received.
2. The save dialog case shows something is kinda messed up with blocking module calls on rpcs3. Merely inserting a sleep_timeout call before entering the input loop even without a separate thread works the same as your method. Probably it interferes with the PPU scheduler and prevents threads from exiting suspended status, I'll think of a clean way around it without spawning another thread since the game already spawns the save dialog thread as a dedicated thread for this very task already and the function is known to be a blocking call. I reckon on ps3 the other threads keep running even if a module call just enters an infinite loop which does not seem to be the case on rpcs3.

In conclusion, the bug in save dialog is not because wait_for has bad behavior, it shouldn't have been used the way it was in the first place. In fact, since the wait there is for a simple variable change, native sleep_for is a better fit imo.

**Comment by jjsat** [comment]
To number 2): I had noticed this behaviour before. Deus Ex HR calls accept() on a thread and before the net rewrite it would freeze the whole game, not only the blocked thread. The scheduler can probably not switch away from a thread while a HLE function is running.

**Comment by scribam** [comment]
@jjsat Do the changes from https://github.com/RPCS3/rpcs3/pull/5244 match more with what you expect?
